rust-sdk/core/src/math/tick_array.rs

```diff
@@ -0,0 +1,252 @@
+use crate::{
+    ErrorCode, TickArrayFacade, TickFacade, INVALID_TICK_INDEX, MAX_TICK_INDEX,
+    TICK_ARRAY_NOT_EVENLY_SPACED, TICK_ARRAY_SIZE, TICK_INDEX_OUT_OF_BOUNDS, TICK_SEQUENCE_EMPTY,
+};
+
+use super::{get_next_initializable_tick_index, get_prev_initializable_tick_index};
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct TickArraySequence<const SIZE: usize> {
+    tick_arrays: [Option<TickArrayFacade>; SIZE],
+    tick_spacing: u16,
+}
+
+impl<const SIZE: usize> TickArraySequence<SIZE> {
+    pub fn new(
+        tick_arrays: [Option<TickArrayFacade>; SIZE],
+        tick_spacing: u16,
+    ) -> Result<Self, ErrorCode> {
+        let mut tick_arrays = tick_arrays;
+        tick_arrays.sort_by_key(start_tick_index);
+
+        if tick_arrays.is_empty() || tick_arrays[0].is_none() {
+            return Err(TICK_SEQUENCE_EMPTY);
+        }
+
+        let required_tick_array_spacing = TICK_ARRAY_SIZE as i32 * tick_spacing as i32;
+        for i in 0..tick_arrays.len() - 1 {
+            let current_start_tick_index = start_tick_index(&tick_arrays[i]);
+            let next_start_tick_index = start_tick_index(&tick_arrays[i + 1]);
+            if next_start_tick_index - current_start_tick_index != required_tick_array_spacing
+                && next_start_tick_index != MAX_TICK_INDEX
+            {
+                return Err(TICK_ARRAY_NOT_EVENLY_SPACED);
+            }
+        }
+
+        Ok(Self {
+            tick_arrays,
+            tick_spacing,
+        })
+    }
+
+    pub fn start_index(&self) -> i32 {
+        start_tick_index(&self.tick_arrays[0])
+    }
+
+    pub fn end_index(&self) -> i32 {
+        let mut last_valid_start_index = self.start_index();
+        for i in 0..self.tick_arrays.len() {
+            if start_tick_index(&self.tick_arrays[i]) != MAX_TICK_INDEX {
+                last_valid_start_index = start_tick_index(&self.tick_arrays[i]);
+            }
+        }
+        last_valid_start_index + TICK_ARRAY_SIZE as i32 * self.tick_spacing as i32
+    }
+
+    pub fn tick(&self, tick_index: i32) -> Result<&TickFacade, ErrorCode> {
+        if (tick_index < self.start_index()) || (tick_index >= self.end_index()) {
+            return Err(TICK_INDEX_OUT_OF_BOUNDS);
+        }
+        if (tick_index % self.tick_spacing as i32) != 0 {
```

Comment:
Yep you are right. It is not possible to use Vec (even when not exposed through wasm) because we want to build the wasm lib without `rust::std`. Will propose a fix